### PR TITLE
[READY] Switch to Codecov for coverage

### PR DESCRIPTION
**Pull request by vheon**
For now this will enable only coverage for the python layer for every build that we make. I will add coverage for the C++ layer later on (I still didn't try it). Just for reference coverage went up to 86.60%.

---
This change is [Reviewable](https://reviewable.io/reviews/valloric/ycmd/524)

### CODE REVIEW

**Comment by micbou** [comment]
We need to replace the Coveralls badge by the Codecov one in README.

<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> #### \[READY\] Switch to Codecov for coverage
> 
> For now this will enable only coverage for the python layer for every build that we make. I will add coverage for the C++ layer later on (I still didn't try it). Just for reference coverage went up to 86.60%.
> 
> 
> 
</details>


---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/tests/utils_test.py, line 287 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/524#-KK-ec5RpJs5CGhXqi_z:-KK-ec5SVtZMVKVnN4tJ:1158180469) ([raw file](https://github.com/valloric/ycmd/blob/cfaac908b7f169bafe61c69be0e5c58013545f92/ycmd/tests/utils_test.py#L287)):*
> ```Python
> 
> 
> @WindowsOnly
> ```

This should be a `@UnixOnly` test. We are checking that the `stdin_windows` parameter is removed on non-Windows platforms.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524#-:-KK-dcef05JSOKclbSY2:237439203)*

**Comment by vheon** [comment]
For that I need this to be merged I think.

<details><summary><i>Previously, micbou wrote…</i></summary>
> We need to replace the Coveralls badge by the Codecov one in README.
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/tests/utils_test.py, line 287 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/524#-KK-ec5RpJs5CGhXqi_z:-KK-uPvZ2aKn4DNRBNok:1424007548) ([raw file](https://github.com/valloric/ycmd/blob/cfaac908b7f169bafe61c69be0e5c58013545f92/ycmd/tests/utils_test.py#L287)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> This should be a `@UnixOnly` test. We are checking that the `stdin_windows` parameter is removed on non-Windows platforms.
</details>
Actually this is testing that we replace `stdin_windows` with `stdin` when its value is `subprocess.PIPE` and that is done only on windows. I actually wanted to write another test for what you're saying but apparently it got lost :unamused: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524#-:-KK-uHe_n-HSWTNr2mbV:1734317923)*

**Comment by vheon** [comment]
Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/tests/utils_test.py, line 287 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/524#-KK-ec5RpJs5CGhXqi_z:-KK-vmzU1sJq8e_ISSXM:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/cfaac908b7f169bafe61c69be0e5c58013545f92/ycmd/tests/utils_test.py#L287)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> Actually this is testing that we replace `stdin_windows` with `stdin` when its value is `subprocess.PIPE` and that is done only on windows. I actually wanted to write another test for what you're saying but apparently it got lost :unamused: 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524)*

**Comment by micbou** [comment]
> For that I need this to be merged I think.

Ok. @Valloric needs to add the repository to Codecov.

:lgtm:



---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524#-:-KK16ZMd28Dsimrg8iag:398782641)*

**Comment by Valloric** [comment]
This is great stuff! Like @micbou said, we need to change the readme badge. Otherwise :lgtm: with minor comments.

I can see [ycmd on codecov.io](https://codecov.io/gh/Valloric/ycmd) just fine; do I need to do something else?

<details><summary><i>Previously, micbou wrote…</i></summary>
> > For that I need this to be merged I think.
> 
> Ok. @Valloric needs to add the repository to Codecov.
> 
> :lgtm:
> 
> 
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[codecov.yml, line 8 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/524#-KK5Mudidf7ycyKJTTBt:-KK5Mudidf7ycyKJTTBu:-73515766) ([raw file](https://github.com/valloric/ycmd/blob/17b323194cf5ae5b05a272542800e58345b08ccd/codecov.yml#L8)):*
> ```YAML
>     precision: 2
>     round: down
>     range: 70...100
> ```

What does this range thing do? Is it "coverage must be in this range to pass?" Ideally we'd have something like "coverage must be greater or equal to before."

Could you put a link to relevant docs here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524#-:-KK5NLWSN6zHltFErUSS:1232213336)*

**Comment by Valloric** [comment]
Badge should be this: [codecov](https://codecov.io/gh/Valloric/ycmd)

Source: `[codecov](https://codecov.io/gh/Valloric/ycmd)`

<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> This is great stuff! Like @micbou said, we need to change the readme badge. Otherwise :lgtm: with minor comments.
> 
> I can see [ycmd on codecov.io](https://codecov.io/gh/Valloric/ycmd) just fine; do I need to do something else?
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524#-:-KK5Od1thfDvZAVEnfOv:1270356247)*

**Comment by vheon** [comment]
I've updated the badge.

<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Badge should be this: [codecov](https://codecov.io/gh/Valloric/ycmd)
> 
> Source: `[codecov](https://codecov.io/gh/Valloric/ycmd)`
</details>


---

Review status: 8 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[codecov.yml, line 8 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/524#-KK5Mudidf7ycyKJTTBt:-KK5WV2_gY4foWrJBAFN:-1807006404) ([raw file](https://github.com/valloric/ycmd/blob/17b323194cf5ae5b05a272542800e58345b08ccd/codecov.yml#L8)):*
It actually defines the range of coverage to map to the range of colors. So in this case if the coverage is 70% the badge would be red and if 100% would be green.

[Here](https://github.com/codecov/support/blob/master/codecov.yml) is the documented default codecov.yml

> Ideally we'd have something like "coverage must be greater or equal to before."

I think is doable but I have to revisit the docs for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524#-:-KK5X74ez6Kd1wx8h6KK:648396845)*

**Comment by vheon** [comment]
Review status: 8 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[codecov.yml, line 8 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/524#-KK5Mudidf7ycyKJTTBt:-KK5Y7jnbnQdSa_hG3z8:1913296911) ([raw file](https://github.com/valloric/ycmd/blob/17b323194cf5ae5b05a272542800e58345b08ccd/codecov.yml#L8)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> It actually defines the range of coverage to map to the range of colors. So in this case if the coverage is 70% the badge would be red and if 100% would be green.
> 
> [Here](https://github.com/codecov/support/blob/master/codecov.yml) is the documented default codecov.yml
> 
> > Ideally we'd have something like "coverage must be greater or equal to before."
> 
> I think is doable but I have to revisit the docs for it.
</details>
Sorry, wrong link. You should find the documentation for all the possible configurations in the settings of the project https://codecov.io/gh/Valloric/ycmd/settings/yaml

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524)*

**Comment by vheon** [comment]
Review status: 8 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[codecov.yml, line 8 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/524#-KK5Mudidf7ycyKJTTBt:-KK5YwkS2aKrorZ8yb2g:1231049908) ([raw file](https://github.com/valloric/ycmd/blob/17b323194cf5ae5b05a272542800e58345b08ccd/codecov.yml#L8)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> Sorry, wrong link. You should find the documentation for all the possible configurations in the settings of the project https://codecov.io/gh/Valloric/ycmd/settings/yaml
</details>
Apparently the "coverage increased/decreased since previous commit" is the default behaviour (which is the same we have with coveralls.io).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524)*

File: codecov.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+codecov:
+  notify:
+    require_ci_to_pass: yes
+
+  coverage:
```
Comment (stevepeak) [inline codecov.yml]: This is under its own node:

```yaml
codecov:
  ...
coverage:
 ...
comment:
 ...
```
> note that `codecov`, `coverage` and `comment` are top lvl nodes.

Thank you!

Comment (vheon) [inline codecov.yml]: Thanks, I'll fix that. I've copied this from https://github.com/codecov/support/blob/master/codecov.yml so you might want to fix that too?

Comment (stevepeak) [inline codecov.yml]: Nice catch! Thank you, Just updated now

**Comment by micbou** [comment]
Bye Coveralls. Hello Codecov.

@homu r+

---

Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/524#-:-KKKwFMN_Ktejd1VLvD5:-1802494803)*

**Pull request merged**